### DISCUSSION TITLE
Flux bootstrap to enterprise gitlab failed

### DISCUSSION THREAD

**Discussion by manugitops** [fluxcd/flux2#4265] [Q&A]
Dears,

flux bootstrap to enterprise gitlab failing with error:

```
✗ failed to get Git repository "https://$GITLAB_URL/$GITLAB_OWNER/test-project": provider error: invalid character '<' looking for beginning of value
```

Command:

```
flux@flux-control-plane:~> flux bootstrap gitlab \
>   --hostname=$GITLAB_URL \
>   --owner=$GITLAB_OWNER \
>   --repository=test-project \
>   --token-auth \
>   --branch=main \
>   --path=clusters/my-cluster \
>   --verbose
Please enter your GitLab personal access token (PAT):
► connecting to $GITLAB_URL
✗ failed to get Git repository "$GITLAB_URL/$GITLAB_OWNER/test-project": provider error: invalid character '<' looking for beginning of value
```

Flux Version: flux version 2.1.0

Gitlab Version: GitLab Enterprise Edition [15.11.13-ee]

Request your kind support. Thank you

**Comment by somtochiama** [comment]
Hey, I just tested bootstrap on `GitLab Enterprise Edition v16.3.2-ee` and it works.
I think this error might be from trying to decode the response from the Gitlab API maybe?

What do you get when you try
```
curl --header "Private-Token: $GITLAB_TOKEN"  $GITLAB_URL/api/v4/projects/<URL encoded path of owner/repository>
```
e.g my project is `flux-tests` and the group is `test-group1`
```
curl --header "Private-Token: $GITLAB_TOKEN"  $GITLAB_URL/api/v4/projects/test-group1%2Fflux-tests (slash is replaced w %2F)
```
See Gitlab docs: https://docs.gitlab.com/ee/api/projects.html#get-single-project

**Reply by manugitops** [reply]
Hi Somtochi,

When I try the suggested command, there is no response (seems empty response)
Cursor just return to the prompt without any error or response.

Thank you

**Reply by somtochiama** [reply]
That's weird, should at least return a 404 if you have the wrong values.

Can you check if you have the wrong group ID? (the group ID is the path in the URL when you visit that group, it sometimes differs from the group name). Does the project already exist? what do you see when you curl for the list of projects in a group
```
curl --header "Private-Token: $GITLAB_TOKEN"  $GITLAB_URL/api/v4/groups/<group-id>/projects
``` 
(Get the group id from the path URL)

**Reply by manugitops** [reply]
Hi Somtochi,

Yes, it seems like gitlab URL is not directly accessible and to simulate that I have used mtls transmitter by creating a pod and service using tls secret.

Now I can see project details in resonse to command "curl --header "Private-Token: $GITLAB_TOKEN"  $GITLAB_URL/api/v4/projects/<URL encoded path of owner/repository>"

but again when I try to bootstap getting another error. If you could please suggest further.

flux@flux-control-plane:~> flux bootstrap gitlab \
> --hostname=$GITLAB_URL \
> --owner=$GITLAB_OWNER \
> --repository=ec-deployments \
> --token-auth \
> --branch=main \
> --path=clusters/my-cluster \
> --verbose
Please enter your GitLab personal access token (PAT):
► connecting to http://gitlab-service.cvc-cd.svc.cluster.local:8080
► cloning branch "main" from Git repository "http://gitlab-service.cvc-cd.svc.cluster.local:8080/manmohan.singh.sorout/ec-deployments.git"
⚠️  clone failure: basic auth cannot be sent over HTTP
⚠️  clone failure: basic auth cannot be sent over HTTP
✗ failed to clone repository: basic auth cannot be sent over HTTP

Thank you.

**Reply by somtochiama** [reply]
You can't use basic auth w HTTP and we don't natively support mTLS for git repositories.
Can you use ssh?

**Reply by manugitops** [reply]
No, getting same basic auth error with --ssh-hostname as well. Is there anything that need to be setup at project/group level.

**Reply by somtochiama** [reply]
Can you try `flux bootstrap git --url=ssh://git@<GITLAB_URL>/<GROUP/REPO>.git --path=clusters/my-cluster` (to use the ssh agent on your machine or you can pass in the ssh key with the `--private-key-file` flag.

**Reply by manugitops** [reply]
Getting connection timeout (used ssh agent).

```
flux@flux-control-plane:~> flux bootstrap git --url=ssh://git@$GITLAB_URL/rcli-temporary/dashboard/test-awesome-project.git --path=clusters/my-cluster
► cloning branch "main" from Git repository "ssh://git@$GITLAB_URL/rcli-temporary/dashboard/test-awesome-project.git"
⚠️  clone failure: unable to clone 'ssh://git@$GITLAB_URL/rcli-temporary/dashboard/test-awesome-project.git': dial tcp 104.18.223.160:22: connect: connection timed out
⚠️  clone failure: unable to clone 'ssh://git@$GITLAB_URL/rcli-temporary/dashboard/test-awesome-project.git': dial tcp 104.18.223.160:22: connect: connection timed out
✗ failed to clone repository: unable to clone 'ssh://git@$GITLAB_URL/rcli-temporary/dashboard/test-awesome-project.git': dial tcp 104.18.223.160:22: connect: connection timed out
```

git clone working fine using the same url "ssh://git@$GITLAB_URL/rcli-temporary/dashboard/test-awesome-project.git"

**Reply by somtochiama** [reply]
Can you try passing in the private key as a flag?

**Comment by VasylR** [comment]
I would suggest to check if you are using cloudflare application firewall in between your host and the gitlab. When you use ssh, your ssh client uses your tls keys to authenticate with cloudflare firewall and another tls key to authenticate with gitlab.
